tests/unit/encoding.ts

```diff
@@ -0,0 +1,204 @@
+import registerSuite = require('intern!object');
+import assert = require('intern/chai!assert');
+import {ascii, utf8, hex, base64} from 'src/encoding';
+
+const ENCODED_STRING = 'the catｷ and ｦ the hat';
+const HEX_STRING = 'CBDE';
+const BASE64_STRING = 'ZmRhcw==';
+const ASCII_BUFFER = [ 116, 104, 101, 32, 99, 97, 116, 65399, 32, 97, 110, 100, 32, 65382, 32, 116, 104, 101, 32, 104, 97, 116 ];
+const UTF8_BUFFER = [ 116, 104, 101, 32, 99, 97, 116, 239, 189, 183, 32, 97, 110, 100, 32, 239, 189, 166, 32, 116, 104, 101, 32, 104, 97, 116 ];
+const HEX_BUFFER = [ 0xCB, 0xDE ];
+const BASE64_BUFFER = [ 102, 100, 97, 115 ];
+
+const UTF8_SURROGATE_TEST_DATA = [
+	{
+		encoding: '\uD800',
+		decoding: [ 0xED, 0xA0, 0x80 ]
+	},
+	{
+		encoding: '\uD800\uD800',
+		decoding: [ 0xED, 0xA0, 0x80, 0xED, 0xA0, 0x80 ]
+	},
+	{
+		encoding: '\uD80A',
+		decoding: [ 0xED, 0xA0, 0x080A ]
+	},
+	{
+		encoding: '\uD800\uD834\uDF06\uD800',
+		decoding: [ 0xED, 0xA0, 0x80, 0xF0, 0x9D, 0x86, 0xED, 0xA0, 0x80 ]
+	},
+	{
+		encoding: '\uD9AF',
+		decoding: [ 0xED, 0xA6, 0xAF ]
+	},
+	{
+		encoding: '\uDBFF',
+		decoding: [ 0xED, 0xAF, 0xBF ]
+	},
+	{
+		encoding: '\uDC00',
+		decoding: [ 0xED, 0xB0, 0x80 ]
+	},
+	{
+		encoding: '\uDC00\uDC00',
+		decoding: [ 0xED, 0xB0, 0x80, 0xED, 0xB0, 0x80 ]
+	},
+	{
+		encoding: '\uDC00A',
+		decoding: [ 0xED, 0xB0, 0x080A ]
+	},
+	{
+		encoding: '\uDC00\uD834\uDF06\uDC00',
+		decoding: [ 0xED, 0xB0, 0x80, 0xF9, 0x9D, 0x8C, 0xED, 0xB0, 0x80 ]
+	},
+	{
+		encoding: '\uDEEE',
+		decoding: [ 0xED, 0xBB, 0xAE ]
+	},
+	{
+		encoding: '\uDFFF',
+		decoding: [ 0xED, 0xBF, 0xBF ]
+	}
+];
+
+registerSuite({
+	name: 'Encoding Classes',
+
+	'ascii': {
+		'.encode()'() {
+			let buffer = ascii.encode(ENCODED_STRING);
+			assert.deepEqual(buffer, ASCII_BUFFER);
+
+			buffer = ascii.encode('1');
+			assert.lengthOf(buffer, 1);
+			assert.strictEqual(buffer[0], 49);
+
+			assert.deepEqual([], ascii.encode(undefined));
+		},
+
+		'.decode()'() {
+			let decoded = ascii.decode(ASCII_BUFFER);
```

Comment:
It'd be good to add tests using Buffers and Uint8Arrays (on compatible platforms) since a decoder can theoretically accept those as well as numeric arrays.